internal/webhook/sentinel/handler.go

```diff
@@ -0,0 +1,187 @@
+		}
+		// replicate sts labels to the /scale subresource object, used for subsequent checks
+		obj.SetLabels(sts.GetLabels())
+	}
+
+	etcdName, hasLabel := obj.GetLabels()[druidv1alpha1.LabelPartOfKey]
+	if !hasLabel {
+		return admission.Allowed(fmt.Sprintf("label %s not found on resource", druidv1alpha1.LabelPartOfKey))
+	}
+
+	etcd := &druidv1alpha1.Etcd{}
+	if err = h.Get(ctx, types.NamespacedName{Name: etcdName, Namespace: req.Namespace}, etcd); err != nil {
+		if apierrors.IsNotFound(err) {
+			return admission.Allowed(fmt.Sprintf("corresponding Etcd %s not found", etcdName))
+		}
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	// allow changes to resources if Etcd has annotation druid.gardener.cloud/disable-resource-protection is set
+	if !etcd.AreManagedResourcesProtected() {
+		return admission.Allowed(fmt.Sprintf("changes allowed, since Etcd %s has annotation %s", etcd.Name, druidv1alpha1.DisableResourceProtectionAnnotation))
+	}
+
+	// allow operations on resources if the Etcd is currently being reconciled, but only by etcd-druid,
+	// and allow exempt service accounts to make changes to resources, but only if the Etcd is not currently being reconciled.
+	if etcd.IsReconciliationInProgress() {
+		if req.UserInfo.Username == h.config.ReconcilerServiceAccount {
+			return admission.Allowed(fmt.Sprintf("ongoing reconciliation of Etcd %s by etcd-druid requires changes to resources", etcd.Name))
+		}
+		return admission.Denied(fmt.Sprintf("no external intervention allowed during ongoing reconciliation of Etcd %s by etcd-druid", etcd.Name))
+	} else {
+		for _, sa := range h.config.ExemptServiceAccounts {
+			if req.UserInfo.Username == sa {
+				return admission.Allowed(fmt.Sprintf("operations on Etcd %s by service account %s is exempt from Sentinel Webhook checks", etcd.Name, sa))
+			}
+		}
+	}
+
+	return admission.Denied(fmt.Sprintf("changes disallowed, since no ongoing processing of Etcd %s by etcd-druid", etcd.Name))
+}
+
+func (h *Handler) decodeRequestObject(req admission.Request, requestGK schema.GroupKind) (client.Object, error) {
+	var (
+		obj client.Object
+		err error
+	)
+	switch requestGK {
+	case
+		corev1.SchemeGroupVersion.WithKind("ServiceAccount").GroupKind(),
+		corev1.SchemeGroupVersion.WithKind("Service").GroupKind(),
+		corev1.SchemeGroupVersion.WithKind("ConfigMap").GroupKind(),
+		rbacv1.SchemeGroupVersion.WithKind("Role").GroupKind(),
+		rbacv1.SchemeGroupVersion.WithKind("RoleBinding").GroupKind(),
+		appsv1.SchemeGroupVersion.WithKind("StatefulSet").GroupKind(),
+		autoscalingv1.SchemeGroupVersion.WithKind("Scale").GroupKind(), // for statefulsets' /scale subresource
+		policyv1.SchemeGroupVersion.WithKind("PodDisruptionBudget").GroupKind(),
+		batchv1.SchemeGroupVersion.WithKind("Job").GroupKind(),
+		coordinationv1.SchemeGroupVersion.WithKind("Lease").GroupKind():
+		obj, err = h.doDecodeRequestObject(req)
+	}
```

Comment:
Handle the default case to return an error for unsupported resource types.

Reply:
If you are suggesting to return error for default case, it may not be the best idea. Because, the caller of `decodeRequestObject()` returns admission response as `Errored` for any errors in decoding the object, whereas if it's an unexpected resource type, it returns `Denied` response. To make this distinction between the two, we return a nil `obj` for an unknown resource type, and the caller handles that. I could add a docstring to document this behavior of the function.

The other way to do this is to return a custom error for default case and check that in the caller, but that may be an overkill, especially because this function is called only once ever.